src/OpenSSL/SSL.py

```diff
@@ -699,12 +699,22 @@ def load_client_ca(self, cafile):
 
     def set_session_id(self, buf):
         """
-        Set the session identifier.  This is needed if you want to do session
-        resumption.
+        Set the session id to *buf* within which a session can be reused for
+        this Context object.  This is needed when doing session resumption,
+        because there is no way for a stored session to know which Context
+        object it is associated with.
+
+        :param bytes buf: The session id.
 
-        :param buf: A Python object that can be safely converted to a string
         :returns: None
         """
+        buf = _text_to_bytes_and_warn("buf", buf)
+        if _lib.SSL_CTX_set_session_id_context(
+                self._context,
+                buf,
+                len(buf),
+        ) != 1:
+            _raise_current_error()
```

Comment:
You can get around the lack of coverage for this case by making a new function:

```python
def pyopenssl_assert(ok):
    if not ok:
        _raise_current_error()
```

then you can change the above to

```python
        res = _lib.SSL_CTX_set_session_id_context(
                self._context,
                buf,
                len(buf),
        )
        _pyopenssl_assert(res != 1)
```

Coverage on `_pyopenssl_assert` is, of course, trivial then. This is the approach we use in cryptography when we have response codes we expect to always be a specific value unless something catastrophic has occurred.